dotCMS/src/main/java/com/dotcms/analytics/AnalyticsAPIImpl.java

```diff
@@ -0,0 +1,244 @@
+     */
+    private void logResponse(final Response response) {
+        if (AnalyticsHelper.isSuccessResponse(response)) {
+            Logger.info(this, "Access Token requested and fetched correctly");
+        } else {
+            Logger.error(this, String.format(
+                "Error requesting access token from IDP server %s due to: %s (status code: %d)",
+                analyticsIdpUrl,
+                response.getStatusInfo().getReasonPhrase(),
+                response.getStatusInfo().getStatusCode()));
+        }
+    }
+
+    /**
+     * Request an access token by sending an HTTP post with analytics app data to defined IDP.
+     *
+     * @param analyticsApp provided analytics app
+     * @return a http response representation
+     */
+    private Response requestAccessToken(final AnalyticsApp analyticsApp) {
+        final Response response = ClientBuilder.newClient()
+            .target(analyticsIdpUrl)
+            .request(MediaType.APPLICATION_FORM_URLENCODED)
+            .header(
+                HttpHeaders.AUTHORIZATION,
+                String.format("Basic %s", analyticsApp.clientIdAndSecret()))
+            .post(Entity.entity("grant_type=client_credentials", MediaType.APPLICATION_JSON_TYPE));
+        logResponse(response);
+        return response;
+    }
+
+    /**
+     * Requests an {@link AccessToken} with associated analytics app data and saves it to be accessible when found.
+     *
+     * @param analyticsApp provided analytics app
+     * @return {@link Optional<AccessToken>} with value when a not expired token is found, otherwise empty\
+     */
+    private AccessToken refreshAccessToken(final AnalyticsApp analyticsApp) throws DotDataException {
+        final Response response = requestAccessToken(analyticsApp);
+
+        return AnalyticsHelper.extractToken(response)
+            .map(accessToken -> {
+                Logger.info(this, "Saving access token to cache");
+                analyticsCache.putAccessToken(accessToken.withIssueDate(new Date()));
+                return accessToken;
+            })
+            .orElseThrow(() -> new DotDataException("Could not extract access token from response, saving aborted"));
+    }
+
+    /**
+     * Request an access token by sending an HTTP post with app's data to defined IDP.
+     *
+     * @param analyticsApp provided analytics app
+     * @return a http response representation
+     */
+    private Response requestAnalyticsKey(final AnalyticsApp analyticsApp) throws DotDataException {
+        final AccessToken accessToken = fetchAccessToken(analyticsApp);
+        final Response response = ClientBuilder.newClient()
+            .target(analyticsApp.getAnalyticsProperties().analyticsConfigUrl())
+            .request(MediaType.APPLICATION_FORM_URLENCODED)
+            .header(HttpHeaders.AUTHORIZATION, String.format("Bearer %s", accessToken.accessToken()))
+            .post(Entity.entity("", MediaType.TEXT_PLAIN_TYPE));
```

Comment:
Super friendly reminder to myself to turn this http method to a GET, per jitsu config script made by @spbolton.